src/Microsoft.AspNet.Razor.Runtime/TagHelpers/DefaultTagHelperContent.cs

```diff
@@ -14,39 +15,41 @@ namespace Microsoft.AspNet.Razor.Runtime.TagHelpers
     /// </summary>
     public class DefaultTagHelperContent : TagHelperContent
     {
-        private readonly BufferEntryCollection _buffer;
+        private BufferedHtmlContent _buffer;
 
-        /// <summary>
-        /// Instantiates a new instance of <see cref="DefaultTagHelperContent"/>.
-        /// </summary>
-        public DefaultTagHelperContent()
-        {
-            _buffer = new BufferEntryCollection();
-        }
-
-        /// <inheritdoc />
-        public override bool IsModified
+        // internal for testing
+        internal BufferedHtmlContent Buffer
         {
             get
             {
-                return _buffer.IsModified;
+                if (_buffer == null)
+                {
+                    _buffer = new BufferedHtmlContent();
+                }
+
+                return _buffer;
             }
         }
 
         /// <inheritdoc />
+        public override bool IsModified => _buffer != null;
+
+        /// <inheritdoc />
+        /// <remarks>Returns <c>true</c> for a cleared <see cref="TagHelperContent"/>.</remarks>
         public override bool IsWhiteSpace
         {
             get
             {
-                foreach (var value in _buffer)
+                if (_buffer == null)
                 {
-                    if (!string.IsNullOrWhiteSpace(value))
-                    {
-                        return false;
-                    }
+                    return true;
                 }
 
-                return true;
+                using (var writer = new EmptyOrWhitespaceWriter())
+                {
+                    WriteTo(writer, HtmlEncoder.Default);
```

Comment:
What we discussed earlier for this and `IsEmpty` would have avoided running through the whole buffer in common cases:
``` c#
    foreach (var entry in _buffer.Entries)
    {
        if (entry == null)
        {
            continue;
        }

        var stringValue = entry as string;
        if (stringValue != null)
        {
            if (!string.IsNullOrWhitespace(stringValue))
            {
                return false;
            }
        }
        else
        {
            ((IHtmlContent)entry).WriteTo(writer, HtmlEncoder.Default);
            if (!writer.IsWhiteSpace)
            {
                return false;
            }
        }
    }
```

Then `return true;` after the `using`.